Add a new 'library' for k8s/openshift providers.

cdrage opened the pull request:
This is the first commit and first initialization into refactoring the
OpenShift provider in order to create a library that is both compatible
with OpenShift as well as Kubernetes.

With this 'library', the *ONLY* thing that is passed in is an object of
the .kube/config configuration. After passing the configuration, you may
make API calls such as .create(object) and .delete(object).

This commit integrates the library as well as converts the current Kubernetes provider from cmd-line (using kubectl) to using the API.

cdrage commented:
#dotests

cdrage commented:
#dotests

cdrage commented:
#dotests

cdrage commented:
#dotests

cdrage commented:
#dotests

cdrage commented:
#dotests

cdrage commented:
Functional tests failing due to: https://github.com/projectatomic/adb-tests/pull/32/files

cdrage commented:
So I've gotten this to work with the Kubernetes provider.

The only thing that's left is the conversion of the OpenShift provider to this new refactor. I've left the `lib/openshift.py` code in there in order to pick-up where I left off when we convert said OpenShift to the new 'library' (with certs and all). I can remove that code if you wish.

Otherwise, this PR is ready for review!

Feel free to give it a spin!

dustymabe commented:
@kadel can we get you to review this?

dustymabe commented:
I feel like we might need to go deeper than `atomicapp/providers/lib/` because lib could be used for more than just the kube/shift library. i.e. we might want libraries for other providers in the future.

Should we add one level: `atomicapp/providers/lib/kubeshift/`

cdrage commented:
kubeshift sounds like a good name, I'll make that change @dustymabe

kadel commented on atomicapp/providers/lib/kubebase.py:
```diff
@@ -0,0 +1,309 @@
+
+        # Lastly, if we have client-certificate and client-key in the .kube/config
+        # we add them to the connection as a cert
+        if ["client-certificate", "client-key"] in self.user.keys():
```
We also need support for `client-certificate-data` and `client-key-data` keys.
If specified those options contain "inlined"  base64 encoded certificate/key.

https://github.com/kubernetes/kubernetes/blob/v1.2.2/pkg/client/unversioned/clientcmd/api/types.go#L83,L84

cdrage replied:
ACK. I'll work on this

dustymabe commented on atomicapp/providers/lib/kubeconfig.py:
```diff
@@ -14,6 +14,66 @@
+    @staticmethod
+    def from_params(api=None, auth=None, ca=None, verify=True):
+        '''
+        Creates a .kube/config based upon the arguments given.
```
`Creates a .kube/config` confused me a bit. This doesn't actually create a file, but rather a data structure with the information in it. How about we use something like: `Creates a dict of .kube/config data based upon the arguments given` ?

cdrage replied:
Updated ^^

dustymabe commented on atomicapp/providers/lib/kubernetes.py:
```diff
@@ -0,0 +1,170 @@
+
+        '''
+
+        # The configuration data passed in will be .kube/config data, so process is accordingly.
```
`process is` ?

dustymabe commented on atomicapp/providers/lib/kubernetes.py:
```diff
@@ -0,0 +1,170 @@
+
+    def _generate_kurl(self, obj, namespace, name=None, params=None):
+        '''
+        Generate the required URL by extracting the 'kind' from the object as well as the namespace.
```
maybe try to keep things under 80 chars

dustymabe commented on tests/units/providers/test_kubernetes_provider.py:
```diff
@@ -53,7 +53,7 @@ def prepare_provider(self, data):
 
     # Check that the provider configuration file exists
-    @mock.patch.object(KubernetesProvider, '_call', mock_provider_call)
+    # @mock.patch.object(KubernetesProvider, '_call', mock_provider_call)
```
No longer needed? Why?

dustymabe commented:
I'm confused right now by `atomicapp/providers/lib/openshift.py`. 
- We don't use this yet do we? Should we add it in a separate PR? 
- It seems to have very similar code to what is in `lib/kubernetes.py`. I see there is a #TODO to refactor. Is the plan to use inheritance here in the future? would seem logical to me since openshift is "theoretically" a superset of openshift.

dustymabe commented on atomicapp/providers/lib/kubernetes.py:
```diff
@@ -0,0 +1,170 @@
+        ns = self.api.request("get", url)
+        return ns['items']
+
+    def _generate_kurl(self, obj, namespace, name=None, params=None):
```
kurl = kind url? 

I wonder if we should make that more explicit

cdrage replied:
Yup. I added that to the comments. We could do "gen_kind_url" instead, would this be better?

dustymabe replied:
yes- i think so

dustymabe replied:
This has not yet been done.

dustymabe commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,234 @@ class KubernetesProvider(Provider):
 
-        Use the first valid value found
+            Default configuration:
+                If no paramaters have been provider, as well as --provider-config,
```
s/provider/provided
s/as well as /as well as no /

dustymabe commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,234 @@ class KubernetesProvider(Provider):
+        # Check that the required parameter is passed. If not, error out.
+        for k in [PROVIDER_API_KEY]:
+            if result[k] is None:
+                msg = "Kubernetes API access: You need to set %s in %s or pass it as a CLI param" % (k, ANSWERS_FILE)
```
very long line.. would be nice if we go back through at some point and try to keep things under 80 chars

cdrage replied:
ACK

dustymabe commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,234 @@ class KubernetesProvider(Provider):
+                self.api = Client(self._from_required_params(), "kubernetes")
+            else:
+                config_location = os.path.join(Utils.getRoot(), Utils.getUserHome().strip('/'), '.kube/config')
+                self.api = Client(KubeConfig.from_file(config_location), "kubernetes")
```
Before this PR (when using kubectl), kubectl would work even if no ~/.kube/config existed. Is there a sensible default we can use to achieve this same level of convenience?

If none of the previous things were existed then we could assume an endpoint (`http://localhost:8080/api`) and then fire off a test connection and if passed then use that as our config. Thoughts?

cdrage replied:
Thanks for pointing this out! Yeah, I'll add this here too. If no .kube/config exists we should try that end-point anyways.

dustymabe commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,234 @@ class KubernetesProvider(Provider):
+
+    def _check_required_params(self):
+        '''
+        This checks to see if any parameters associated to the Kubernetes provider are passed
```
are we checking to make sure required paramters are passed and then erroring if not? 

What are required parameters for kubernetes? I don't think there are the same requirements as there are for openshift so we might have to change this up a bit. I tried to make a providerconfig and get it to work but there was too much information that was required.

cdrage replied:
I'll add a doc for this. Only the API end-point is required for the kubernetes provider.

dustymabe replied:
I don't really see anything about this function that has been updated. I still don't understand why it is trying to verify the existence of 4 values when only 1 is required

cdrage replied:
@dustymabe this checks to see if any CLI params or answers have been passed. does that make sense? so when you pass `--provider-auth --provider-api --provider-cert, etc.` it will create a generate ~/.kube/config using that information. Similar to how the configuration is generated with the `openshift.py` provider.

dustymabe commented on atomicapp/providers/lib/kubebase.py:
```diff
@@ -0,0 +1,309 @@
+
+        # Gather the user, context and cluster details of .kube/config based upon the current_context
+        kubeconfig_context = self._contexts()[self.current_context]
+        kubeconfig_user = kubeconfig_context['user']
```
user isn't required.. this config works for me using kubectl:

```
apiVersion: v1
clusters:
- cluster:
    server: "http://localhost:8080"
  name: foo
contexts:
- context:
    cluster: "foo"
  name: foocont
current-context: foocont
kind: Config
preferences: {}
users: []
```

cdrage replied:
ACK. I added 'user' due to picking up the "token" information. You are correct that it is not needed. I will update the PR for this edge-case.

cdrage commented:
#dotests

kadel commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,234 @@ class KubernetesProvider(Provider):
 
-        :return: str -- Resource name / identity
+    '''
+    This is DEPRECATED and not needed anymore as we check the /resource URL of the kubernetes api against the artifact
```
Can we remove this?

cdrage commented:
@dustymabe @kadel @rtnpro 

Hey all,

I've updated it based upon @dustymabe 's comments.. 

In regards to your comments, I've fixed the edge-case with the 'user' not being defined in the .kube/config class.

Same goes for when there is no .kube/config. Atomic App will generate the required api credentials.

I've moved it to `kubeshift` as well.

@kadel I haven't added the client-cert yet, but I will soon as I continue to work on this.

I'll also add a doc @dustymabe on what the required params are (just like

dustymabe commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,243 @@ class KubernetesProvider(Provider):
-        if self.dryrun:
-            # Testing env does not have kubectl in it
-            return "/usr/bin/kubectl"
+            config_location = os.path.join(Utils.getRoot(), Utils.getUserHome().strip('/'), '.kube/config')
```
what do you think about calling this `default_config` or `default_config_location` or something like that?

cdrage replied:
default_config_loc maybe? I'll change it.

dustymabe commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,243 @@ class KubernetesProvider(Provider):
+
+            Default configuration:
+                If no paramaters have been provided, as well as no --provider-config,
+                use the default .kube/config location and load the associated configuration.
```
I think these last two could use some more clarity. How about: 

```
            .kube/config:
                If no config file or params are provided by user then try to find and
                use a config file at the default location.

            no config at all:
                If no .kube/config file can be found then try to connect to the default
                unauthenticated http://localhost:8080/api end-point.
```

cdrage replied:
ACK

dustymabe commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,243 @@ class KubernetesProvider(Provider):
+                self.api = Client(KubeConfig.from_file(self.config_file), "kubernetes")
+            elif self._check_required_params():
+                self.api = Client(self._from_required_params(), "kubernetes")
+            elif not os.path.isfile(config_location):
```
I think it would be more intuitive if we make this one `elif os.path.isfile()` and move the localhost:8080 stuff to the `else:`

cdrage replied:
ACK

dustymabe commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,243 @@ class KubernetesProvider(Provider):
-    def init(self):
-        self.namespace = "default"
+    # Essential provider parameters
+    provider_api = "http://localhost:8080"
```
the code below that uses this "default" value would actually make more sense if we set `localhost:8080` there instead of here. I would just set this to `None` unless there is a technical reason why we shouldn't. 

also, one could make an argument that this should be a constant??

cdrage replied:
ACK and updated

dustymabe commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,243 @@ class KubernetesProvider(Provider):
+                self.api = Client(KubeConfig.from_file(config_location), "kubernetes")
+
+            # Check if the namespace that the app is being deployed to is available
+            self._check_namespaces()
```
is this supposed to raise an exception? It currently does not.

cdrage replied:
It should, ACK.

dustymabe commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,243 @@ class KubernetesProvider(Provider):
+
+        self._process_artifacts()
 
         if not self.dryrun:
```
maybe we should just make this: `return if self.dryrun` and then we can unindent the rest of the function?

dustymabe commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,243 @@ class KubernetesProvider(Provider):
+            # Check if the namespace that the app is being deployed to is available
+            self._check_namespaces()
+
+    def _check_required_params(self):
```
btw this is a great example of a function that can be implemented for kubernetes and openshift - since they will probably have different requirements of what is required. openshift could inherit from a library and then "override" this function with one of its own.

dustymabe commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,243 @@ class KubernetesProvider(Provider):
+            result[k] = self.config.get(k)
+
+        # Check that the required parameter is passed. If not, error out.
+        for k in [PROVIDER_API_KEY]:
```
i see now where we are checking that only provider api key is set.. and it is the only one required.. we need to have some coordination between `_from_required_params` and `_check_required_params` since they both "check required params"

cdrage replied:
This is taken from the code that was used in `providers/openshift.py` in regards to checking the CLI params. I can change this to an if statement instead.

What are your thoughts? Refactor it into another function?

dustymabe replied:
yeah somehow consolidate the two.. here is one way it could be done: 
[foofile.txt](https://github.com/projectatomic/atomicapp/files/285318/foofile.txt)

dustymabe commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,243 @@ class KubernetesProvider(Provider):
+        self.k8s_artifacts[kind].append(data)
+
+    '''
+    This is DEPRECATED and not needed anymore as we check the /resource URL of the kubernetes api against the artifact
```
we'll eventually remove this - but thanks for leaving it in.. 

can you elaborate? what code are we doing now that we weren't in the past?  i.e. where do we "check the /resource URL" and why does doing that make this no longer needed. 

@tkral may be interested in this

dustymabe replied:
Can you remove this function in a 2nd commit?

dustymabe replied:
Would still like @tkral's opinion.

dustymabe commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,243 @@ class KubernetesProvider(Provider):
-
-            if kind in ["ReplicationController", "rc", "replicationcontroller"]:
-                self._scale_replicas(path, replicas=0)
+        logger.info("Deploying to Kubernetes")
```
`Undeploying from`

dustymabe commented:
I notice we removed persistent storage code - with the 'rescoping' of nulecule does it make sense that we don't try to do persistent storage in the future?

dustymabe commented on atomicapp/providers/lib/kubeshift/client.py:
```diff
@@ -0,0 +1,57 @@
+        else:
+            raise KubeClientError("No provider by that name.")
+
+    # Create an object using it's respective API
```
s/it's/its

dustymabe commented on atomicapp/providers/lib/kubeshift/client.py:
```diff
@@ -0,0 +1,57 @@
+    def create(self, obj, namespace="default"):
+        self.connection.create(obj, namespace)
+
+    # Delete an object using it's respective API
```
s/it's/its

dustymabe commented on atomicapp/providers/lib/kubeshift/kubebase.py:
```diff
@@ -0,0 +1,310 @@
+                # Disable the 'InsecureRequestWarning' notifications.
+                # As per: https://github.com/kennethreitz/requests/issues/2214
+                # Instead make a large one-time noticable warning instead
+                requests.packages.urllib3.disable_warnings()
```
+1

dustymabe commented on atomicapp/providers/lib/kubeshift/kubebase.py:
```diff
@@ -0,0 +1,310 @@
+                # As per: https://github.com/kennethreitz/requests/issues/2214
+                # Instead make a large one-time noticable warning instead
+                requests.packages.urllib3.disable_warnings()
+                logger.warning("CAUTION: TLS verification has been DISABLED")
```
++1

dustymabe commented:
With @rtnpro's Config PR I wonder if we shouldn't try to disambiguate a bit with our use of the `config` word within this PR. I wonder if we should use `kconfig` instead. Or something similar.

dustymabe commented on atomicapp/providers/kubernetes.py:
```diff
@@ -38,227 +44,243 @@ class KubernetesProvider(Provider):
+        for k in result.keys():
+            result[k] = self.config.get(k)
+
+        if bool([a for a in result.values() if a is None]):
```
Too fancy.. could this be an alternative:

```
>>> values = [False, False, "foo"]
>>> if None in values:
...     print "noneinvalues"
... 
>>> values = [False, False, None]
>>> if None in values:
...     print "noneinvalues"
... 
noneinvalues
```

cdrage replied:
But I like fancy! ACK xD

dustymabe replied:
I just don't like it when the alternative (much easier to read) version right there for the taking :) 

I was only making a suggestion so we can leave it like it is, but do you agree that one is easier to read than the other?

dustymabe commented on tests/units/providers/test_kubernetes_provider.py:
```diff
@@ -53,7 +53,9 @@ def prepare_provider(self, data):
-    @mock.patch.object(KubernetesProvider, '_call', mock_provider_call)
+    # We no longer use the mock call as _call no longer exists within the
+    # kubernetes provider. Thus we test the configuration exists
+    # @mock.patch.object(KubernetesProvider, '_call', mock_provider_call)
```
then we don't need mock_provider_call function either.

dustymabe commented:
We moved `kubeconfig.py` so let's update the openshift provider:

```
  File "/opt/atomicapp/atomicapp/providers/openshift.py", line 42, in <module>
    from atomicapp.providers.lib.kubeconfig import KubeConfig
ImportError: No module named kubeconfig
```

cdrage commented:
Hey @dustymabe 
I updated the code again with your comments.

Could we keep the 'fancy' code? Seems a lot cleaner.

Yes in regards to the persistent storage. Since the scope has changed and the fact that persistent storage was not used (as much as we hoped) I purposely left it out.

dustymabe commented:
> Could we keep the 'fancy' code? Seems a lot cleaner.
> 
I commented on that in  the other comment. The short answer is yes. 

> Yes in regards to the persistent storage. Since the scope has changed and the fact that persistent storage was not used (as much as we hoped) I purposely left it out.

We need to spell this out more explicitly - i.e. separate PR to remove the storage stuff and an announcement. Let's make sure we think through this first as well. Leave the PR as it is right now with respect to the storage stuff though.

dustymabe commented:
> With @rtnpro's Config PR I wonder if we shouldn't try to disambiguate a bit with our use of the config word within this PR. I wonder if we should use kconfig instead. Or something similar.

What did you think of this comment ^^ ?

cdrage commented:
I agree in regards to the config terminology. It'll be nice to separate the two. Have it use `kconfig` here instead.

I'll also update this with the combine of the two functions as you pointed out in foo.txt

dustymabe commented:
> I'll also update this with the combine of the two functions as you pointed out in foo.txt

you like the name ?? :)

cdrage commented:
Hey @dustymabe thank you for being so up-to-date with the reviewing.

So that's it for now in terms of code! I've updated based on your comments and added a few entries for logger.debug output in the `kubernetes.py` provider.

I'm going to do some last-minute testing (in terms of ssl gen + using certs) to make sure it's all 100%.

cdrage commented:
#dotests

dustymabe commented:
There is a `atomicapp/providers/.kubernetes.py.swn` file in here - not sure where it came from but needs to be removed.

cdrage commented:
@dustymabe ewwwww. WHY VIM WHY. Removed.

dustymabe commented:
> I agree in regards to the config terminology. It'll be nice to separate the two. Have it use kconfig here instead.

I don't think you updated this

cdrage commented:
@dustymabe @kadel @rtnpro 

I just tested against https://github.com/pires/kubernetes-vagrant-coreos-cluster and it seems to work correctly :+1: could you guys test against your environments for any errors / issues?

cdrage commented:
@dustymabe sorry, was a bit confused on what you meant. is this regarding any of the references to `self.config`?  Since that inherits from the `config` of the Provider class?

So the best way to separate the two would be to simply do `self.kconfig = self.config` on startup and use kconfig?

dustymabe commented:
> @dustymabe sorry, was a bit confused on what you meant. is this regarding any of the references to self.config? Since that inherits from the config of the Provider class?
> 
> So the best way to separate the two would be to simply do self.kconfig = self.config on startup and use kconfig?

```
config (object): An object of the .kube/config configuration
```

I think anything that is referencing an object of the `.kube/config` configuration should be known as kconfig. This would be an attempt to disambiguate between kconfig and config (the class that rtnpro is writing).

cdrage commented:
@dustymabe ahhhh, that's what you meant! I was getting confused between @rtnpro 's refactor and `.kube/config`, I'll change the code to reflect that :)

kadel commented:
I tested this on [Kubernetes running on AWS](http://kubernetes.io/docs/getting-started-guides/aws/)

When authenticating using token, everything works fine.

When using `client-certificate`, `client-key` it fails
```
▶ atomicapp-dev stop . --provider=kubernetes --provider-config=/home/tomas/.kube/config --namespace=default --verbose                           
atomicapp-dev:1: command not found:  
INFO   :: - cli/main.py ::    _  _             _      _
INFO   :: - cli/main.py ::   /_\| |_ ___ _ __ (_)__  /_\  _ __ _ __   Version:  0.5.2
INFO   :: - cli/main.py ::  / _ \  _/ _ \ '  \| / _|/ _ \| '_ \ '_ \  Nulecule: 0.0.2
INFO   :: - cli/main.py :: /_/ \_\__\___/_|_|_|_\__/_/ \_\ .__/ .__/  Mode:     Stop
INFO   :: - cli/main.py ::                                |_|  |_|
DEBUG  :: - cli/main.py :: Final parsed cmdline: stop . --provider=kubernetes --provider-config=/home/tomas/.kube/config --namespace=default --verbose
DEBUG  :: - nulecule/main.py :: NuleculeManager init app_path: .
DEBUG  :: - nulecule/main.py :: NuleculeManager init image: None
DEBUG  :: - utils.py :: Loading answers from file: ./answers.conf
DEBUG  :: - utils.py :: Loading answers from file: ./answers.conf.gen
DEBUG  :: - providers/kubernetes.py :: Given config: {u'provider-tlsverify': False, u'provider-config': u'/home/tomas/.kube/config', u'provider': u'kubernetes', u'image': u'centos/httpd', u'hostport': 80, u'namespace': u'default'}
INFO   :: - providers/kubernetes.py :: Using namespace default
DEBUG  :: - providers/kubernetes.py :: Processing artifact: artifacts/kubernetes/.hello-apache-pod.json
DEBUG  :: - utils.py :: Finding the users home directory
DEBUG  :: - utils.py :: Running as user tomas. Using home directory /home/tomas for configuration data
DEBUG  :: - providers/kubernetes.py :: Provider configuration provided
ERROR  :: - cli/main.py :: 'token'
Traceback (most recent call last):
  File "/home/tomas/dev/github/kadel/atomicapp/atomicapp/cli/main.py", line 130, in cli_func_exec
    cli_func(cli_func_args)
  File "/home/tomas/dev/github/kadel/atomicapp/atomicapp/cli/main.py", line 97, in cli_stop
    nm.stop(**argdict)
  File "/home/tomas/dev/github/kadel/atomicapp/atomicapp/nulecule/main.py", line 349, in stop
    self.nulecule.stop(cli_provider, dryrun)
  File "/home/tomas/dev/github/kadel/atomicapp/atomicapp/nulecule/base.py", line 223, in stop
    component.stop(provider_key, dryrun)
  File "/home/tomas/dev/github/kadel/atomicapp/atomicapp/nulecule/base.py", line 351, in stop
    provider.init()
  File "/home/tomas/dev/github/kadel/atomicapp/atomicapp/providers/kubernetes.py", line 101, in init
    self.api = Client(KubeConfig.from_file(self.config_file), "kubernetes")
  File "/home/tomas/dev/github/kadel/atomicapp/atomicapp/providers/lib/kubeshift/client.py", line 42, in __init__
    self.connection = KubeKubernetesClient(config)
  File "/home/tomas/dev/github/kadel/atomicapp/atomicapp/providers/lib/kubeshift/kubernetes.py", line 41, in __init__
    self.api = KubeBase(config)
  File "/home/tomas/dev/github/kadel/atomicapp/atomicapp/providers/lib/kubeshift/kubebase.py", line 69, in __init__
    self.token = self.user['token']
KeyError: 'token'
```

cdrage commented:
@kadel 
Could you post your .kube/config (with your API credentials balnked out and whatever) I think this is an edge-case with user being set with no token available.

kadel commented:
@cdrage I don't think that this is edge case, this is how authentication for `system:admin` user in OpenShift is setup by default (using  `client-certificate-data` and `client-key-data`).

If you have config with both client-certificate/key and token for one user, you have to decide which one to use, it doesn't make sense to use both, even if in this case it is technically possible.

Examples of `.kube/config` with  different auth methods:
- client-{certificate,key} - https://gist.github.com/kadel/b2d3df8962a2d2feabc1c5b29de82b55
- client-{certificate,key}-data - https://gist.github.com/kadel/29df1fee245627ed90e53b44cdb7c2e7
- token - https://gist.github.com/kadel/1089d0906a997da017f5a4bc4f39bb0e
- username:password (BasicAuth) - https://gist.github.com/kadel/2a5c4bb74ab733688ab1eaf4976080b7

cdrage commented:
@kadel thank you for posting them! I'm going to go through each-one and figure out how we can work this in.

cdrage commented:
@kadel how did we do `certificate-authority-data` in the openshift.py provider? or was this intended to be added in the cert commit for openshift?

kadel commented:
@cdrage yes it was part of old PR with cert auth. Here is example how we handled that https://github.com/kadel/atomicapp/blob/bdc3a2bd3cc953ea1436392367a3f92a1b532b8d/atomicapp/providers/lib/kubeconfig.py#L116,L119
Temporray files was becuase requests library accepts certs only as files, or I didn't figure out how to do this differently

kadel commented on atomicapp/providers/lib/kubeshift/kubebase.py:
```diff
@@ -0,0 +1,331 @@
+        '''
+
+        # If it starts with /, we assume it's a filename so we just return that.
+        if data.startswith('/'):
```
it can be related path to `.kube/config` and than it is not starting with `/`

kadel commented on atomicapp/providers/lib/kubeshift/kubebase.py:
```diff
@@ -0,0 +1,331 @@
+        # If it's data, we assume it's a base64 certificate and we decode and write to a tmp file
+        else:
+            with tempfile.NamedTemporaryFile(delete=False) as f:
+                f.write(base64.b64decode(data))
```
certificates and keys in files are not base64 encoded

cdrage replied:
Thanks for this @kadel :) I've updated the PR to reflect what parts should be decoded and what shouldn't.

I'll have to add the other `-data` keys in a separate PR

cdrage commented:
@kadel updated the PR with your feedback on the `-data` files. In particular, it's in `kubebase.py` in relation to the `cert_file` declarations around the requests library.

kadel commented:
Sorry, but It is not working like this :-(
You are still expecting that under one key in `.kube/config` can be either base64 encoded string or filename. But this is not right. 
For example CA certificate can be in `certificate-authority` (than it is path to file) or in `certificate-authority-data` than it is base64 encoded string.

kadel commented:
There are still some missing features to properly handle Kubernetes config file.
But we can add those in another PR.
 - support for relative paths to certificates and keys
 - support for *-data keys (`certificate-authority-data`, `client-certificate-data`, `key-certificate-data`)

:+1:  LGTM
Awesome work @cdrage :sparkles:

cdrage commented:
@kadel 

Thanks man! :dancers: 

I'll add the `-data` as a separate PR as well as the support for relative paths (will have to use the `utils.py` functions we had before to find the proper path, due to /host being passed, etc.)

cdrage commented:
@dustymabe I think this may be good to go! I've updated kubebase to reflect the wording change on self.config to self.kubeconfig

This may be ready for mergin' :+1:

cdrage commented:
#dotests

dustymabe commented:
@cdrage - I haven't looked at this but from what I reviewed a week ago everything looked good. Since Tomas gave +1 I'm +1. Merge if you can get all tests passing.

cdrage commented:
WOOO! Kubernetes tests pass! :tada: 

Merging away :)

cdrage commented:
Thanks everyone for helping me out with this! It's been exciting to finally get this in.

dustymabe commented:
haha openshift tests failed - thoughts about that?

cdrage commented:
@dustymabe false alarm, tests passed locally for OpenShift on my system :)

Seems the previous issue is occuring again with the latest update to OpenShift:
```

ERROR  :: - cli/main.py :: 403 {u'status': u'Failure', u'kind': u'Status', u'code': 403, u'apiVersion': u'v1', u'reason': u'Forbidden', u'details': {u'kind': u'pods'}, u'message': u'User "system:anonymous" cannot create pods in project "foo"', u'metadata': {}}
```

cdrage commented:
Let me test one more thing ^^ updated my OpenShift image and it failed.

cdrage commented:
@dustymabe should be fine, no need to revert or anything. i knew I didn't touch the openshift integration (didn't change any underlying code)

I've opened up https://github.com/projectatomic/adb-tests/pull/33 to fix this.

cdrage commented:
#dotests

khrisrichardson commented on atomicapp/providers/lib/kubeshift/kubernetes.py:
```diff
@@ -0,0 +1,171 @@
+            raise KubeKubernetesError("Kubernetes API URL does not include HTTP or HTTPS")
+
+        # Gather what end-points we will be using
+        self.k8s_api = urljoin(url, "api/v1/")
```
Are there plans to support the other groups of APIS accessible via the /apis endpoint (e.g. - apps, autoscaling, batch, extensions, policy, etc.)?

cdrage replied:
@khrisrichardson We hope so. Any changes to https://github.com/projectatomic/nulecule will reflect in atomic app in terms of providing support for autoscaling, batch commands, extensions, policies, etc. If the spec changes we'll reflect that as features in Atomic App :)

khrisrichardson replied:
Thanks, @cdrage 

Is there a timeline in mind, because by restricting to api/v1, the behavior may be a regression for some?

```
ERROR  :: No kind by that name: Deployment
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/atomicapp-0.5.2-py2.7.egg/atomicapp/cli/main.py", line 130, in cli_func_exec
    cli_func(cli_func_args)
  File "/usr/local/lib/python2.7/dist-packages/atomicapp-0.5.2-py2.7.egg/atomicapp/cli/main.py", line 84, in cli_run
    nm.run(**argdict)
  File "/usr/local/lib/python2.7/dist-packages/atomicapp-0.5.2-py2.7.egg/atomicapp/nulecule/main.py", line 320, in run
    nulecule.run(cli_provider, dryrun)
  File "/usr/local/lib/python2.7/dist-packages/atomicapp-0.5.2-py2.7.egg/atomicapp/nulecule/base.py", line 224, in run
    provider.run()
  File "/usr/local/lib/python2.7/dist-packages/atomicapp-0.5.2-py2.7.egg/atomicapp/providers/kubernetes.py", line 268, in run
    self.api.create(artifact, self.namespace)
  File "/usr/local/lib/python2.7/dist-packages/atomicapp-0.5.2-py2.7.egg/atomicapp/providers/lib/kubeshift/client.py", line 49, in create
    self.connection.create(obj, namespace)
  File "/usr/local/lib/python2.7/dist-packages/atomicapp-0.5.2-py2.7.egg/atomicapp/providers/lib/kubeshift/kubernetes.py", line 62, in create
    kind, url = self._generate_kurl(obj, namespace)
  File "/usr/local/lib/python2.7/dist-packages/atomicapp-0.5.2-py2.7.egg/atomicapp/providers/lib/kubeshift/kubernetes.py", line 142, in _generate_kurl
    raise KubeKubernetesError("No kind by that name: %s" % kind)
KubeKubernetesError: No kind by that name: Deployment
```

I was testing kubernetes v1.3.0-alpha.5, and prior to pulling this commit, I was able to create objects belonging to the extensions/v1beta1 API, such as DaemonSet and Deployment.

cdrage replied:
@khrisrichardson Ah! That's what you meant. I'm going to open up an issue for this.

khrisrichardson replied:
I have a patch to support API groups. Will fork and create PR.
